### PR TITLE
[tests] utils: refactor type-hint signatures.

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Refactoring

### Purpose
- Addresses some discussion from https://github.com/sphinx-doc/sphinx/pull/12109#discussion_r1527196512

### Detail
- Consolidate to provision of a single `http_server` utility method, with `tls_enabled` as a boolean flag.
- Rework a few type hints, attempting to make them easier to understand.

### Relates
- Borrows some ideas from #12143.

### CODE REVIEW

File: tests/utils.py
Comment (picnixz) [inline tests/utils.py]:
```suggestion
def http_server(handler: type[BaseRequestHandler], *, tls_enabled: bool = False) -> Iterator[HttpServerThread]:
```

File: tests/utils.py
Comment (picnixz) [inline tests/utils.py]: At the top, just after future imports.

**Review by danieleades — APPROVED** [review]

File: tests/utils.py
Comment (picnixz) [inline tests/utils.py:4]:
```suggestion
```

Comment (picnixz) [inline tests/utils.py]: I think there should be 1 line after and not two (but I'm not sure with the diff)

**Review by danieleades — APPROVED** [review]
LGTM

**Comment by jayaddison** [comment]
Thanks @danieleades!

**Review by picnixz — APPROVED** [review]

File: tests/utils.py
Comment (chrisjsewell) [inline tests/utils.py]:
```suggestion
__all__ = ("http_server",)
```

just a pet-peeve 😄 , but can we use tuples for `__all__`, after all they should be immutable no?

File: tests/utils.py
Comment (chrisjsewell) [reply tests/utils.py]: could always remove this file from the format exclude list 😄

File: tests/utils.py
Comment (jayaddison) [reply tests/utils.py]: _Should_ be, yep.. can you point to some documentation that says that `tuple` is OK as well as `list`?  (I couldn't find any, initially)

**Review by chrisjsewell — APPROVED** [review]
a few minor nitpicks, but otherwise all good thanks

File: tests/utils.py
Comment (danieleades) [reply tests/utils.py]: I can't argue with your logic, but I've never seen a tuple in the wild

Comment (chrisjsewell) [reply tests/utils.py]: mypy (https://github.com/python/mypy/pull/1640) and pylance recognise any sequence of strings

> I can't argue with your logic, but I've never seen a tuple in the wild

I always use them 😉  https://github.com/executablebooks/markdown-it-py/blob/1064835b30325e857252f5b67b34e622261723ca/markdown_it/__init__.py#L2

Comment (chrisjsewell) [reply tests/utils.py]: I guess this is the "formal" documentation: https://github.com/python/typing/blob/main/docs/source/libraries.rst#library-interface-public-and-private-symbols

Comment (picnixz) [reply tests/utils.py]: For `__all__` I tend to use a list because I don't want to put the comma if I only have a single element. It's only by laziness that I'm doing this though.

Comment (jayaddison) [reply tests/utils.py]: > I guess this is the "formal" documentation: https://github.com/python/typing/blob/main/docs/source/libraries.rst#library-interface-public-and-private-symbols

Thanks, perfect.  I've derived a permalinked from that, https://github.com/python/typing/blob/9a39406fa45fbfa76eb5d35afb7863f5f727ad87/docs/source/libraries.rst?plain=1#L108-L119 - and will use tuples and reference that.

> For `__all__` I tend to use a list because I don't want to put the comma if I only have a single element. It's only by laziness that I'm doing this though.

Agreed it looks a little unusual.  However the immutability benefit does seem genuine, so I'll apply this.

File: tests/utils.py
Comment (jayaddison) [reply tests/utils.py]: If you don't mind waiting until this is merged, we could do that as part of #12146?

**Comment by chrisjsewell** [comment]
shall we merge this @jayaddison?

**Comment by jayaddison** [comment]
@chrisjsewell yep, this is ready to merge from my perspective :+1:

**Comment by jayaddison** [comment]
Thank you!

**Pull request merged**
